// use-local-storage-state/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const react_1 = require("react");
function parseJSON(value) {
    return value === 'undefined'
        ? undefined
        :
            JSON.parse(value !== null && value !== void 0 ? value : '');
}
const data = {};
const storage = {
    get(key, defaultValue) {
        var _a;
        try {
            return (_a = data[key]) !== null && _a !== void 0 ? _a : parseJSON(localStorage.getItem(key));
        }
        catch (_b) {
            return defaultValue;
        }
    },
    set(key, value) {
        try {
            localStorage.setItem(key, JSON.stringify(value));
            data[key] = undefined;
            return true;
        }
        catch (_a) {
            data[key] = value;
            return false;
        }
    },
    remove(key) {
        data[key] = undefined;
        try {
            localStorage.removeItem(key);
        }
        catch (_a) { }
    },
};
const initializedStorageKeys = new Set();
function useLocalStorageState(key, defaultValue) {
    const [defaultValueState] = react_1.useState(() => {
        const isCallable = (value) => typeof value === 'function';
        return isCallable(defaultValue) ? defaultValue() : defaultValue;
    });
    const [state, setState] = react_1.useState(() => {
        return {
            value: storage.get(key, defaultValueState),
            isPersistent: (() => {
                if (typeof window === 'undefined') {
                    return true;
                }
                try {
                    localStorage.setItem('--use-local-storage-state--', 'dummy');
                    localStorage.removeItem('--use-local-storage-state--');
                    return true;
                }
                catch (_a) {
                    return false;
                }
            })(),
        };
    });
    const updateValue = react_1.useMemo(() => {
        const fn = (newValue) => {
            const isCallable = (value) => typeof value === 'function';
            if (isCallable(newValue)) {
                setState((state) => ({
                    value: newValue(state.value),
                    isPersistent: storage.set(key, newValue),
                }));
            }
            else {
                setState({
                    value: newValue,
                    isPersistent: storage.set(key, newValue),
                });
            }
        };
        fn.reset = () => {
            storage.remove(key);
            setState({
                value: defaultValueState,
                isPersistent: state.isPersistent,
            });
        };
        return fn;
    }, [key, defaultValueState]);
    react_1.useEffect(() => {
        if (initializedStorageKeys.has(key)) {
            throw new Error(`Multiple instances of useLocalStorageState() has been initialized with the same key. ` +
                `Use createLocalStorageStateHook() instead. ` +
                `Example implementation: ` +
                `https://github.com/astoilkov/use-local-storage-state#create-local-storage-state-hook-example`);
        }
        else {
            initializedStorageKeys.add(key);
        }
        return () => void initializedStorageKeys.delete(key);
    }, []);
    react_1.useEffect(() => {
        const onStorage = (e) => {
            if (e.storageArea === localStorage && e.key === key) {
                setState({
                    value: storage.get(key, defaultValueState),
                    isPersistent: true,
                });
            }
        };
        window.addEventListener('storage', onStorage);
        return () => window.removeEventListener('storage', onStorage);
    }, [defaultValueState]);
    return [state.value, updateValue, state.isPersistent];
}
exports.default = useLocalStorageState;
function createLocalStorageStateHook(key, defaultValue) {
    const updates = [];
    return function useLocalStorageStateHook() {
        const [value, setValue, isPersistent] = useLocalStorageState(key, defaultValue);
        const updateValue = react_1.useMemo(() => {
            const fn = (newValue) => {
                for (const update of updates) {
                    update(newValue);
                }
            };
            fn.reset = () => {
                for (const update of updates) {
                    update.reset();
                }
            };
            return fn;
        }, []);
        react_1.useEffect(() => {
            initializedStorageKeys.delete(key);
        }, []);
        react_1.useEffect(() => {
            updates.push(setValue);
            return () => void updates.splice(updates.indexOf(setValue), 1);
        }, [setValue]);
        return [value, updateValue, isPersistent];
    };
}
exports.createLocalStorageStateHook = createLocalStorageStateHook;
